### PR TITLE
MS Word with UIA: use custom extensions to support line number,  layout column number and section number

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
None.

### Summary of the issue:
When supporting Microsoft word via UI automation, up to now there was no way to report line numbers, layout column numbers or section numbers. This was because these concepts do not exist in the UI Automation standard.
However, Microsoft recently added custom patterns to UI automation, as a part of its Custom Extensions work, which included custom properties and custom annotations.
Microsoft word has now added its own custom pattern which allows fetching information from a text range, such as line number, layout column number and section number.

### Description of how this pull request fixes the issue:
Using the microsoft/microsoft-ui-uiautomation Remote Operations library, fetch line number, layout column number and section number in Microsoft word.
Note that the Remote Operations library is only supported in Windows 11 and above, and the custom pattern in Microsoft Word is only available as of build 16.0.14725.

A part from adding support for these specific attributes, this PR also acts as a first commit to add integrations with the UI Automation Remote Operations library. Now that we are building and including this library, there is much more code that can eventually be remoted for a massive speedup.

### Testing strategy:
* Opened a document in Microsoft word
* Turned on reporting of line numbers in NVDA's document formatting settings
* Moved up and down the document with the arrow keys, ensuring that NVDA spoke the current line number before the content of the line.
* Ensured report pages was turned on in NVDA's document formatting settings.
* Set up the document in 2-column layout, and inserted a column break in the content.
* used the  arrow keys to move between the two columns, ensuring that the column number is announced as it changes. 

### Known issues with pull request:
None known.

### Change log entries:
New features
Changes
Bug fixes
For Developers

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

### CODE REVIEW

File: .gitmodules
Comment (codeofdusk) [inline .gitmodules:35]: Are you intending to refer to your fork here, and not microsoft/microsoft-ui-uiautomation?

**Comment by michaelDCurran** [comment]
At this stage yes. I'm still waiting on some PRs to get merged.

File: source/UIAHandler/remote.py
Comment (lukaszgo1) [inline source/UIAHandler/remote.py:4]: This list of contributors is incorrect - probably it was mistakenly copied from some other file
```suggestion
# Copyright (C) 2021-2022 NV Access Limited
```

File: .gitmodules
Comment (michaelDCurran) [reply .gitmodules:35]: Yes, there is still one PR I'm waiting to be merged to upstream.

File: nvdaHelper/UIARemote/UIARemote.cpp
Comment (feerrenrut) [inline nvdaHelper/UIARemote/UIARemote.cpp:15]: Could you add a comment with an example warning and where you see it? Makes it easier to work out when/if this is no longer required.

**Review by feerrenrut — CHANGES_REQUESTED** [review]
There are a few build warnings introduced, these may be outside your control, but it would be good if they were silenced:
```
"C:\work\repo\nvda\PR\include\microsoft-ui-uiautomation\src\uiAutomation\microsoft.ui.uiautomation\Microsoft.UI.UIAutomation.vcxproj" (Build target) (1) ->
(PrepareForBuild target) ->
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppBuild.targets(499,5): warning MSB8004: Output Directory does not end with a trailing slash.  This build instance will add the slash as it is required to allow proper evaluation of the Output Directory. [C:\work\repo\nvda\PR\include\microsoft-ui-uiautomation\src\uiAutomation\microsoft.ui.uiautomation\Microsoft.UI.UIAutomation.vcxproj]
```

File: nvdaHelper/microsoft-ui-uiautomation/sconscript
Comment (michaelDCurran) [reply nvdaHelper/microsoft-ui-uiautomation/sconscript:45]: As these globs are only to find source files for the sake of checking if they are out of date (and a build is required), including too many is worse than missing some. Note that this list is never passed to a compiler for actual building. I would much prefer to keep the glob lines, rather than list out all 30 or more source files, which may very well get added to and we wouldn't know... it just wouldn't rebuild if those new ones changed.

File: nvdaHelper/UIARemote/UIARemote.cpp
Comment (feerrenrut) [inline nvdaHelper/UIARemote/UIARemote.cpp:123]: This seems like something that belongs in a utility function. It is hard to determine correctness at a glance, and easy to make a mistake. 

I think there is a chance that `wideBuf ` is may not be null terminated here.  Using `MultiByteToWideChar` will include return a size which includes the null termination if `cbMultiByte` is `-1`, however this uses the `size()` from `basic_string` which does not count the null terminating character.

An alternative to explicitly dynamically allocating `wchar` array to store the converted string is to use a `std::vector<wchar_t`. One advantage of the vector approach is the easy of initializing the memory eg:
``` c++
std::vector<wchar_t> wideBuf (wideLen, L'\0');
```
This however sacrifices speed for correctness, given this is exception handling I'd say it is a fair tradeoff.

To keep the current heap array approach, just add 1 to `wideLen` when allocating, then after the second call to `MultiByteToWideChar` set `wideBuf[wideLen]=L'\0'`

```suggestion
		auto wideBuf = std::make_unique<wchar_t[]>(1 + wideLen); // add space for a null char.
		MultiByteToWideChar(CP_UTF8, 0, what.c_str(), what.length(), wideBuf.get(), wideLen);
		wideBuf[wideLen] = L'\0';
		LOG_ERROR(L"msWord_getCustomAttributeValue exception: "<<(wideBuf.get()));
```

File: nvdaHelper/UIARemote/remoteLog.h
Comment (feerrenrut) [inline nvdaHelper/UIARemote/remoteLog.h:64]: See other exception handling comments. I'd extract this out into a shared utility.

File: nvdaHelper/readme.md
Comment (feerrenrut) [inline nvdaHelper/readme.md:27]: I don't think there is any C++/CX code left now.

File: nvdaHelper/readme.md
Comment (feerrenrut) [inline nvdaHelper/readme.md:30]: missing cap
```suggestion
This dll is loaded directly in to NVDA. It provides the following features:

```

File: nvdaHelper/readme.md
Comment (feerrenrut) [inline nvdaHelper/readme.md:43]: Missing cap
```suggestion
This dll is loaded by NVDA, providing utility functions that perform certain tasks or batch procedures on Microsoft UI Automation elements.

```

File: nvdaHelper/readme.md
Comment (feerrenrut) [inline nvdaHelper/readme.md:47]: typo
```suggestion
As a dependency of UIARemote.dll, the open source [Microsoft-UI-UIAutomation Remote Operations library](https://github.com/microsoft/microsoft-ui-uiautomation)  is also built.

```

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: nvdaHelper/microsoft-ui-uiautomation/sconscript
Comment (feerrenrut) [reply nvdaHelper/microsoft-ui-uiautomation/sconscript:17]: Could you action this? What are the outputs of this `sconscript`. Looking at the additions to the readme, an explanation doesn't really belong there, but a doc string at the top of this file would help to know what it is responsible for.

File: nvdaHelper/microsoft-ui-uiautomation/sconscript
Comment (feerrenrut) [reply nvdaHelper/microsoft-ui-uiautomation/sconscript:45]: Ok that is fine. The `msbuild` in the action looks like these source files are used for building though?

File: nvdaHelper/microsoft-ui-uiautomation/sconscript
Comment (michaelDCurran) [reply nvdaHelper/microsoft-ui-uiautomation/sconscript:17]: Yep, added a top-level docstring now.

File: nvdaHelper/microsoft-ui-uiautomation/sconscript
Comment (michaelDCurran) [reply nvdaHelper/microsoft-ui-uiautomation/sconscript:60]: the msbuild command references `$SOURCE`. This variable only ever contains the first source file. You'd need to use `$SOURCES` to have all of them appear in the command.

File: nvdaHelper/UIARemote/remoteLog.h
Comment (feerrenrut) [inline nvdaHelper/UIARemote/remoteLog.h:22]: Looks like C++ standard library has you covered here. No need to use the Windows C API: https://stackoverflow.com/a/38383389

``` C++
#include <iostream>
#include <string>
#include <locale>
#include <codecvt>
std::string str = "your string in utf8";
std::wstring_convert<std::codecvt_utf8_utf16<char16_t>> converter;
std::wstring wstr = converter.from_bytes(str);
```


For reference, since C++17 `basic_string` returns a non-const pointer via `data()` you can use it as a buffer for string types to save a copy. In this case you don't need to explicitly reserve space for the trailing null char.
```suggestion
	std::wstring wideBuf (wideLen, L'\0');
	MultiByteToWideChar(CP_UTF8, 0, from.c_str(), from.length(), wideBuf.data(), wideLen);
	return wideBuf;

```

**Review by feerrenrut — COMMENTED** [review]
I'll leave it up to you about this, I think the C++ standard library approach is less error prone and sets a good example.

File: nvdaHelper/UIARemote/UIARemote.cpp
Comment (feerrenrut) [reply nvdaHelper/UIARemote/UIARemote.cpp:85]: I don't see how that can be true (yet!). It makes sense that this would happen in the `Resolve` call, but I think that in this case it can only result in two cross platform calls. It won't know about the second `scope.if` until the lambda is executed, which must happen locally. It can't execute the lambda until it has the result of the first condition.

I suspect that within `Resolve` when an additional `scope.if` is called from a handler, it is recursively resolved immediately.

Comment (feerrenrut) [reply nvdaHelper/UIARemote/UIARemote.cpp:85]: After chatting about this. The missing piece is that both the body and the else lambda's are executed immediately with the goal of constructing remote ops instructions. This means the instruction tree produced can be run in its entirety on in the remote process, ensuring that all logic is executed with a single process call. The log message are recorded on both sides of the condition, but only emitted based on the result of the condition.

Note: this means that modifying any local (non-remote ops type) variable within these lambdas is likely a bug. 
We'll have to watch out for local variables sneaking into these lambdas.

**Review by feerrenrut — APPROVED** [review]

**Comment by feerrenrut** [comment]
The test failure looks unrelated. I have kicked off the build again.
Old build link: https://ci.appveyor.com/project/NVAccess/nvda/builds/42592615/tests

**Pull request merged**
